### ISSUE TITLE
Make it focus without bringing window to the front

### ISSUE THREAD

**Issue by samholmes** [sbmpost/AutoRaise#47]
Similar to ctrl+opt+click, I'd like to config this tool to focus a window without bringing it to the front of all other windows. The primary use-case for this tool for me is to focus on a window for quick access to keyboard shortcuts for that window. The focus without raise feature could be instant or customized to a specific delay.

If I want to bring the window to the front, then it would be nice to make this tool bring the window to the front after 500ms of no cursor motion. This way I can move my cursor anywhere without raising windows but focusing instantly on windows. Only after stopping my cursor on one window will the window be raised. This is the ideal UX I'd like to achieve. Is this doable?

**Comment by dreeves** [comment]
Beautifully said and I think you're absolutely right that this is the ideal behavior. It's how it works in Linux too! (They call it focus-follows-mouse + auto-raise, where the focus-follows-mouse part is always instant and the auto-raise delay is configurable, including the option to not auto-raise at all.)

Unfortunately @sbmpost has looked into it and concluded that Apple just doesn't expose the ability to focus without auto-raising.

If anyone thinks of some clever way to somehow make it happen, that would be amazing.

PS: Dup of #7

**Comment by sbmpost** [comment]
@dreeves thanks for covering this one :-) And indeed it is true this cannot be done ATM.

**Comment by sbmpost** [comment]
@dreeves @samholmes I suggest we leave this issue open, so others can take note of this as well. Thoughts?

**Comment by dreeves** [comment]
Sounds smart! I'd even advocate for adding notes here about why exactly it's impossible and to solicit ideas on conceivable ways to work around Apple's limitations. I'm guessing it's baked deep into the OS and is truly hopeless but who knows!

**Comment by BenJackGill** [comment]
This describes the desired functionality very succicntly: 

> focus-follows-mouse part is always instant and the auto-raise delay is configurable, including the option to not auto-raise at all

Perhaphs a work around could be to instantly auto-foucs, but put on a large custom delay on the auto-raise. The delay could be so large that infact it operates similar to having no auto-raise at all. For example, a Delay setting of a few hours or days that only affects the auto-focus part and not the focus-follows-mouse part.

Is that possible?

I tried changing the Delay setting in the config file to a large number (i.e. 100000000000) to test this theory, but it had no effect.

Is the current Delay setting supposed to affect both the focus-follows-mouse and auto-raise at the same time? Because if you could seperate out those Delays (one delay setting for focus-follows-mouse, and one delay setting for auto-raise) then this workaround might suffice.

**Comment by sbmpost** [comment]
@BenJackGill 
The problem is that OSX needs to be instructed to raise the window in order to focus it. As such the two always come together. There is no other way to focus a window.

**Comment by fredngo** [comment]
Ugh, this is exactly the behaviour I'd like to see, super bummed that it seems not possible right now!

**Comment by dreeves** [comment]
Ooh, exciting! I switched to branch 7-47-focus-without-raise-experimental and built it (confirming that it's v2.8 running) and that works to have focus-follows-mouse without auto-raise! But what I'm really pining for is immediate focus-follows-mouse with 500ms-delayed auto-raise. 

So tantalizingly close now! <3

**Comment by sbmpost** [comment]
@dreeves ok! Thanks for the feedback. Out of curiosity: what is the difference between raising a window after 500ms or focusing it first while the mouse is hovering over it? I cannot imagine someone to move the mouse while simulatiously using the keyboard to access that window.

**Comment by Laaarge** [comment]
I think I see what he wants: with 500ms autoraise, you wouldn't have any problems with windows raising while moving the mouse ect, and you also (with immediate focus follows mouse) not have any issues of your first keypresses being registered in the wrong window.

I think that's a good idea, although I'll try focus instantly follows mouse(without raise), it may be enough for my usecase.

**Comment by sbmpost** [comment]
@Laaarge 
> and you also (with immediate focus follows mouse) not have any issues of your first keypresses being registered in the wrong window.

I see, yes that could be it. Thanks.

**Comment by dreeves** [comment]
Exactly right. The default 40ms delay kinda drives me crazy when I'm a bit too slow with moving the mouse across my various displays. But any longer and it would be too annoying waiting for the focus to change. Immediate focus change with delayed autoraise elegantly gets the best of both worlds!

**Comment by Laaarge** [comment]
So I've been using this and it is great, it fits my needs much better than traditional AutoRaise.
However I had 2 problems:

- When getting a system popup window (for example: deleting something, closing a running terminal, macOS asking for a password) the popup would flicker, (and be harder 
- When trying to access the small popup menus of brave extensions, they would disappear as soon as I move the mouse after clicking (to get the menu open), it would close the menu… not Ideal

But:
for better or for worse, I have the Noobest solutions of all:
I removed lines 138-154 (because verbose logged `>>>>>>>>psn equals<<<<<<<<<`)
And that has fixed both of my problems… Lol

I do have a remaining problem: 
In both of those situations, verbose still logs in a loop `Activate` and `focus without raise, $window info$`,
which I don't think is the intended behavior

PS PS: I'm not pushing this since it's such an ugly fix, but if someone wants to take a bite at it, I think this is a great starting place (and I'm going to look into doing this properly)

PS PS PS: I notice that AutoRaise's %CPU (in Activity Monitor) is much higher when moving (fast) around in 1 window, then switching focus very frequently (and even higher than when flickering popup menu) (haven't tested it on main branch though)

Thanks a lot for this App!

**Comment by sbmpost** [comment]
@Laaarge 
Thank you for you feedback, always appreciated :-)

I pushed a new version that removes the logging. Also some changes in the way it finds the window under the mouse.

**Comment by Laaarge** [comment]
'twas nothing, thanks for the quick reaction (and work!) : )
>These lines handle the situation where you have 1 single app with multiple windows and you move the mouse between those windows.

Oh ok (strangely, it didn't run into an issue with that, but testing it now, yeah it didn't switch windows within 1 app)

>at least the popup flickering should be gone

Yup confirmed

>The price however is that any window that is smaller than the window below it, will keep the focus. I am not sure if this is better/worse than the previous situation.

I'll keep an eye on that, but it's something that definitely won't be frequent, if it even happens to me: I can't think of cases where that would cause a problem… but that situation definitely exists :'(

> Not sure about the brave extensions. For these it would be interesting to see what the window title of the popup is

it's a bit weird, it doesn't give anything else than the tab name, and I saw a few different behaviors for this, but it seems now that it works without problems the latest commit (popups only disappearing if the browser loses & regains focus).
verbose only logged me this when I was having issues… 
but I don't think it's worth looking into, as it seems fixed (though if you want to, do ask me, as there is more info)
```
… 18:04:29.518 AutoRaise[…] focus without raise, New Tab - Brave
… 18:04:29.518 AutoRaise[…] >>>>>>>>psn equals<<<<<<<<<
```

**Comment by sbmpost** [comment]
@dreeves @Laaarge @samholmes 
So it would seem I have implemented the holy grail then ;-)
Check out the latest version of [this branch](https://github.com/sbmpost/AutoRaise/tree/7-47-focus-without-raise-experimental) and run this:

    make clean && make && ./AutoRaise -delay 25

Of course you can play with the delay to suit your needs.

**Comment by dreeves** [comment]
This is amazing! Nice work! It's working beautifully for me. Now I'm wondering if there's any possible use case for it to ever _not_ work this way...

**Comment by sbmpost** [comment]
@dreeves @Laaarge @samholmes @BenJackGill @fredngo 
I have done a few more fixes and tweaks so you might want to update again.
If nothing else too serious pops up, I will now leave this branch alone for a while.

I suggest to keep this issue open for future reference and additional feedback.
Thanks everyone :-)

**Comment by dreeves** [comment]
I'm on the latest as of now and it's smooth as butter. I'm really, really happy with this!

I'm also really curious to hear if there's [some argument or use case](https://xkcd.com/1172/) for the focus-follows-mouse part to ever be non-immediate.

(Maybe one reason is that you don't trust it to keep working in future macOS versions? That would be a tragedy.)

PS: Note that this GitHub issue is a slight misnomer now (#7 characterizes it more fully) but that's fine to keep the discussion here.

PPS: We know there are some people like @Laaarge and @samholmes and @fredngo who want focus-follows-mouse without autoraise at all. Maybe it would make sense to overload the -delay option like so:

* delay = -1 (aka infinity) means no autoraise, only focus-follows-mouse (which is instant)
* delay = 0 means focus-follows-mouse and autoraise both happen instantly, no delay
* delay = positive number means focus-follows-mouse is instant with subsequent autoraise happening after the given amount of delay

I.e., focus-follows-mouse could always be instant and the delay for autoraise, including whether it happens at all, could be specified by the -delay parameter.

**Comment by sbmpost** [comment]
@dreeves 
> I'm also really curious to hear if there's [some argument or use case](https://xkcd.com/1172/) for the focus-follows-mouse part to ever be non-immediate.
> (Maybe one reason is that you don't trust it to keep working in future macOS versions? That would be a tragedy.)
Exactly that :-)

> Maybe it would make sense to overload the -delay option
Currently it works like this:

- delay = 0: no autoraise, only focus-follows-mouse (which is instant)
- delay = 1: focus-follows-mouse and autoraise both happen instantly, *almost* no delay
- delay = positive number means focus-follows-mouse is instant with subsequent autoraise happening after the given amount of delay

So very close to what you are proposing actually ;-)

**Comment by samholmes** [comment]
I'm in the `-delay 0` club because I think if I want to raise a window, I will click it. How do I bundle this command `AutoRaise -delay 0` into an app bundle?

**Comment by sbmpost** [comment]
@samholmes 
> How do I bundle this command AutoRaise -delay 0 into an app bundle?

You probably don't ;-). To pass arguments to the app bundle version you can use a configuation file (see the README).

**Comment by fredngo** [comment]
> * delay = -1 (aka infinity) means no autoraise, only focus-follows-mouse (which is instant)

I'm in the `delay = -1` club for sure. I don't want it to ever auto-raise (despite the name of this app, lol). I just want it to focus-follow-mouse immediately. I guess this is the behavior I just got used to working on old Unix machines. 😆

**Comment by fredngo** [comment]
> * delay = 0: no autoraise, only focus-follows-mouse (which is instant)

Oops, just saw this -- I guess I'm in the `delay = 0` club instead then.

**Comment by fredngo** [comment]
> I am observing that the menu bar at the top of the screen is changing apps as the mouse pointer is still in motion and crosses other open apps. This is no good. How is one supposed to access the bar when there are other windows in the way

Yup, finally got around to building and running the experimental setup. Works well except for this one problem!

**Comment by fredngo** [comment]
By the way, I went ahead and created a homebrew formula for the experimental branch which makes it way easier to compile/install/etc for most folks!

https://github.com/fredngo/homebrew-autoraise-experimental

The HOWTO is in the README, but here's a synopsis.

Don't forget to uninstall the original home-brew formula if you already had it installed before:

```
brew uninstall autoraise
```

Update your `~/.config/AutoRaise/config` file to disable auto raise:

```
delay=0
```

Then:

```zsh
brew tap fredngo/autoraise-experimental
brew install --HEAD fredngo/autoraise-experimental/autoraise
brew services start autoraise
```

**Comment by sbmpost** [comment]
@fredngo 
> By the way, I went ahead and created a homebrew formula for the experimental branch which makes it way easier to compile/install/etc for most folks!

That is cool 👍 

@git-rz 
> detect when the pointer stops moving before focus

The patch below will accomplish this. However I am not sure if like the behaviour of it. 

    -                        if (raiseTimes) { raiseTimes--; }
    +if (raiseTimes && !mouseMoved) { raiseTimes--; }
    +//                        if (raiseTimes) { raiseTimes--; }
                             else { raiseTimes = 3; }
     
                             if (focusFirst) {
    +if (!mouseMoved) {
                                 OSStatus error = GetProcessForPID(mouseWindow_pid, &mouseWindow_psn);
                                 if (!error) {
                                     window_manager_focus_window_without_raise(&mouseWindow_psn,
    @@ -865,6 +867,7 @@ void onTick() {
                                     _lastFocusedWindow = _mouseWindow;
                                     lastFocusedWindow_pid = mouseWindow_pid;
                                     if (delayCount) { [workspaceWatcher windowFocused: _lastFocusedWindow]; }
    +}

**Comment by git-rz** [comment]
@sbmpost , At long last!! This is perfect!! What is not to like?

Admittedly I haven't poked around on menus and other places where autoraise/autofocus can go wrong for apps that were not built with it in mind. But this implementation proves once and for all that autofocus and the Mac Menu Bar can co-exist, no problem. I can only assume that the Menu bar is the reason why apple made this so difficult to implement. But here we are - a brilliant implementation that finally works.

**Comment by sbmpost** [comment]
@git-rz 
> What is not to like?

Guess I got used to raised windows by now, but your enthusiasm encourages me to try this out for a while :-)

> I can only assume that the Menu bar is the reason why apple made this so difficult to implement.

Around 900 lines of code, yeah I guess this is/was quite a challenge to get right.

> a brilliant implementation that finally works.

Thank you!

@dreeves @fredngo @samholmes @BenJackGill @Laaarge 
Would be great if you can tell me what for you guys the preferred way of working is. With or without the above patch? So I know whether or not to permanently include it in this branch.

**Comment by Laaarge** [comment]
> Would be great if you can tell me what for you guys the preferred way of working is. With or without the above patch?

I'm a bit torn… on one hand, the menu bar is an issue with the default behavior, but I see 2 downsides (they're actually pretty minor, but hey we're way past necessary here :)
1. Feels less responsive: with how obvious MacOs's focus is, it feels weird that the window focuses only after stopping the mouse
2. Needs precise movements to even work if there is another display on the other side of the menu bar

I think we're in a not so simple situation where the right way is going to depend on people's use cases…
Personally, I don't quite know what I prefer… 
I'm going to try the patch, as I have had issues where I needed to drag a window to the menu bar, and I think I'm going to get used to the "delay". But it is a win some, lose some situation.
I'm going to let the other weigh in.

**Comment by fredngo** [comment]
> @dreeves @fredngo @samholmes @BenJackGill @Laaarge
> Would be great if you can tell me what for you guys the preferred way of working is. With or without the above patch? So I know whether or not to permanently include it in this branch.

Since the branch is experimental anyway; if you go ahead and put it into the branch I'd be happy to try it out. No sense holding anything back from the experimental branch.

**Comment by sbmpost** [comment]
@fredngo 
done

**Comment by sbmpost** [comment]
@git-rz 
After my lastest push to this branch it goes like this:

    1 Focus follow mouse ? -> delay = 0 (disables raise)
    2 Autoraise ? -> delay = 1 (1 means no delay at all)
    3 Autoraise delay [ms] -> delay > 1

In all these cases, the mouse should first come to a full stop before the delay is considered. I am not sure if that is the correct behaviour. Perhaps for option 3, a mouse stop should not be required? Or perhaps this should be made configurable? If on the other hand option 3 has become obsolete because of the mouse stop solution, we might also get rid of the delay completely.

**Comment by sbmpost** [comment]
@Laaarge 
Thanks for your feedback. I especially agree with the responsiveness part. Been using the mouse stop solution today, but I go faster then the focus/raise kicking in. I am considering the following:

If raise enabled, don't wait for a mouse stop. If raise is disabled (focus only), demand a mouse stop

**Comment by sbmpost** [comment]
After some more thought I have come to the conclusion it is probably best to make the "mouse stop solution" a configurable setting in this branch. Let me not make the same mistake as Apple and pretend to know what is best for people ;-)

**Comment by git-rz** [comment]
Something @samholmes requested in the original issue, is two delays actually.

I'm still of the mind that delays were a work-around due to lack of mouse-stop logic, and it may be possible to do-away with them totally, but if you are going for full configurability, and especially for his originally requested use case, where he wants focus first, then raise if he hovers too long:

```
[--mouse]         use mouse stop logic      [default: true]
[--delay x]       auto-raise after x ms     [default: 0]
[--focus-delay x] auto-focus after x ms     [default: 0]
[--no-raise]      focus follows mouse only  [default: false]
```

Raise implies focus, so logically the focus delay should only ever be less than the raise delay.

**Comment by dreeves** [comment]
> configurable setting [...] pretend to know what is best for people

I'm not sure yet but I actually think you should default to pretending to know what's best for people and at least wait for a specific use case before making something a setting. The maintenance burden can explode combinatorially as settings proliferate! (I've made the case for that philosophy at <https://blog.beeminder.com/choices>.) @git-rz's argument for always using the mouse-stop criterion sounds persuasive to me (but I haven't tried it yet). I'm still loving instant focus-follows-mouse with 500ms-delayed autoraise.

> delays were a work-around due to lack of mouse-stop logic

Fascinating theory! Now I'm curious to try a version with no delays, just focus+autoraise the moment the mouse stops. If that works then maybe only a single binary setting is needed: when the mouse stops should only the focus change or should autoraise also happen?

**Comment by git-rz** [comment]
> Fascinating theory! 

But my guess falls apart for the people who want immediate focus but delayed raise. It just isn't my preference, so I forgot about it.

Looking at Windows Xmouse, as well as some Linux implementations, it seems like the focus-delay is not a thing (always immediate), whereas the raise-delay is an option.

**Comment by sbmpost** [comment]
@git-rz 
A lot of things to think about :) You provide good arguments to make mouse stop the default. Other things to consider are simplicity (too many options might become confusing) and what people got used to. Plus there is the complication that Apple may change the way to do "focus only" as they please as it is completely undocumented (one of the reasons I am keeping this in an experimental branch). 

I am going to need some time figuring out what is best. Until then, I hope you guys can tweak things in a way that is working for you.

**Comment by git-rz** [comment]
Recently I've been using a tiling window manager (Amethyst), so no windows overlap anyway :) But the mouse-stop logic will help even in the normal auto-raise case. I am hoping others can test, to prove this. You may be on the cutting edge of window activation strategy here. There are a very large number of bugs related to bad interactions of sloppy focus with new UI elements, or subsequent to keyboard navigation, an errant window being activated (or deactivated). This strategy, if not totally new, is certainly working for me, and seems to eliminate the potential for many of those types of defects.  

Thanks for this great utility :D

**Comment by samholmes** [comment]
Because focus-follows-mouse is instant and the app menu changes based on what is focused, I feel like having a delay after mouse is still of say 100 ms would be useful to avoid accidentally focusing on windows while you're mouse is moving towards the menu.

**Comment by git-rz** [comment]
@samholmes , please try the branch with `-delay 0`. You might be surprised that the delay isn't needed at all anymore, not even for menu bar access when crossing apps.

But I agree with you: there still is a use case for a separate option for focusDelay. It would be there for people who may have trouble moving the mouse smoothly. Amputees and the elderly come to mind. Or wireless mice suffering interference or low batteries. The delay would provide forgiveness in those scenarios. 

Regarding the unit: it is currently 20 ms.

**Comment by sbmpost** [comment]
@dreeves @samholmes @fredngo @BenJackGill @git-rz @Laaarge 

> Fascinating theory! Now I'm curious to try a version with no delays, just focus+autoraise the moment the mouse stops. If that works then maybe only a single binary setting is needed: when the mouse stops should only the focus change or should autoraise also happen?

I have pushed another change to this branch that allows you to easily test the best solution by toggling these variables on top of the AutoRaise.mm file:

    focusFirst = true/false
    mouseStop = true/false

I think together with -delay (0=no-raise, 1=no-delay, 2=20ms, 3=40ms) this covers all preferences one could have. I am considering to backport the mouseStop flag to the master branch. By doing so this branch will be just about adding focus first functionality.

**Comment by git-rz** [comment]
I'm testing the new code now. Only feedback so far is that the default value for the `mouseStop` is `false`, and that is opposite the value needed if we need more people testing the new behavior on the experimental branch. ;)

**Comment by fredngo** [comment]
> I'm testing the new code now. Only feedback so far is that the default value for the `mouseStop` is `false`, and that is opposite the value needed if we need more people testing the new behavior on the experimental branch. ;)

Ya I was just about to comment that after recompiling and trying the latest it doesn't seem to wait for the mouse to stop, I guess that explains it!

**Comment by sbmpost** [comment]
@fredngo 
Command line also works, however, it could be you were running 2 instances at the same time. The second time you tried, you probably stopped all instances correctly.

**Comment by fredngo** [comment]
Pretty sure I stopped and restarted everything correctly; but it's ok for now. I'll wait till things stabilize and verify things with the homebrew formula. I'm not as good with dealing with bleeding edge stuff.

But, I'm happy to report that with the new "waiting for mouse to stop" feature, so far things seem to be feeling very good!

If it's not too much to ask, I would even like it if it would wait a *little bit longer* to detect that a mouse has truly stopped!

**Comment by sbmpost** [comment]
> If it's not too much to ask, I would even like it if it would wait a little bit longer to detect that a mouse has truly stopped!
@fredngo you might play a bit with the POLLING_MS value for that. For instance set it to 30 or 40ms or so.

**Comment by fredngo** [comment]
OK, I figured out why sometimes the new build is not working. You have to go in to Security & Privacy and **delete** the old permission, **then** launch the new binary, and have it request permissions again, and approve that. Wow.

<img width="780" alt="Screen Shot 2022-05-05 at 12 46 41 PM" src="https://user-images.githubusercontent.com/11791/166972547-90525933-ec22-4d15-9e43-692e8bfcaf9b.png">

**Comment by sbmpost** [comment]
@fredngo 
This is actually mentioned in the README under troubleshooting ;-)

**Comment by git-rz** [comment]
I found a bug with when -delay=0. It does not matter if mouseStop is true or false. Changing the POLLING_MS to 90 makes it less frequent, but still easily reproducible.

The issue is that the right-click menu disappears almost immediately when right-clicking on the rightmost tab in iTerm2, such that the context menu mostly appears outside the main window:

![image](https://user-images.githubusercontent.com/53906153/167057159-85c2df37-4d2c-4c86-8a7e-6562c8f48400.png)

Changing delay to 1 or more makes the issue go away.

**Comment by fredngo** [comment]
> @fredngo you might play a bit with the POLLING_MS value for that. For instance set it to 50 or 80ms or so. This will also lower cpu usage

This seems to help! I raised mine to 75 and it seems to behave well for my mousing pattern. Now just need a way to configure it from the config file :)

Thank you for all your hard work!

**Comment by git-rz** [comment]
Latest update solves the corner case, so thanks! 

If you fear the fix risked other breakage, where exactly should I focus testing?

**Comment by sbmpost** [comment]
@dreeves 
> It would be amazing if that only applied to autoraise, not focus-follows-mouse.

done

@git-rz 
> If you fear the fix risked other breakage, where exactly should I focus testing?

I am no longer focusing/raising a window that doesn't have the main window attribute. Do all applications set their main window attribute correctly? Wouldn't dare to tell. Is there always a main window? Wouldn't know either ;-). It seems to work so far however.

@fredngo 
> Now just need a way to configure it from the config file :)

fixing all these corner cases is more CPU intensive so I had to increase the default to 40ms. Making this configurable with a certain minimum could be useful, yet complicates things as it would be yet another option.

**Comment by git-rz** [comment]
> Only if the currently focused window has the main window attribute, will I allow the mouse window to be focused. Do all applications set their main window attribute correctly? Wouldn't dare to tell. Is there always a main window? Wouldn't know either ;-). It seems to work so far however.

Good to know. I was perplexed when the pop-up disappeared because the mouse wasn't moving when I right-clicked. I assumed that all focus or unfocus logic would only be evaluated after a mouse-stop. Another way to fix it might be to not count a click, or right-click, as a mouse-stop. But that might have another set of issues.

thanks again!

**Comment by sbmpost** [comment]
@git-rz 
The mouse wasn't moving indeed, but....
As soon as the mouse starts moving, it is not hovering inside the popup yet. Then, when you slowly move your mouse towards the popup, it is very likely a mouse stop happens before you get there and it will focus the window surrounding it again. I have just pushed another refactoring based on the "MainWindow" attribute. This will hopefully be the last change for a while ;-)

**Comment by git-rz** [comment]
> Then, when you slowly move your mouse towards the popup, it is very likely a mouse stop happens before you get there.

Yes, the code was triggered, but I still don't know why. I can guarantee that I did not actually stop.. It was only a very small number of pixels over to get to the menu.

In any event, the latest code is working great so far. I'm very much looking forward to mouse-stop making it into mainline :)

**Comment by sbmpost** [comment]
@git-rz @Laaarge @samholmes @dreeves @fredngo @BenJackGill 

I have added the -mouseStop flag so you can now say this:

    ./AutoRaise -delay 0 -warpX 0.5 -warpY 0.1 -scale 2.5 -mouseStop true

This flag will be added to master soon.

**Comment by fredngo** [comment]
> I have added the -mouseStop flag so you can now say this:

Awesome -- and is it specifiable in the config file also?

**Comment by koekeishiya** [comment]
@sbmpost FYI, there are some applications in which this does not work:

https://github.com/koekeishiya/yabai/issues/397

**Comment by sbmpost** [comment]
@koekeishiya 
Thank you for the heads up. Yes, "focus only" under OSX proves to be a real pain. Thanks to your reverse engineering efforts we at least have it working somewhat, which is cool.

**Comment by fredngo** [comment]
Unfortunately I have to report a bug -- Launchpad is buggy when the autoraise is running. When I open Launchpad it sometimes just goes away as soon as I move the mouse. Or sometimes Launchpad stays on the screen for a while and then goes away as I'm trying to move toward an icon.

I'm not sure if it's because of this new branch or if the behavior exists in the main branch also, but I can't seem to isolate the behavior any more than that. I'll report more as I observe the behavior, if I can.

**Comment by sbmpost** [comment]
@fredngo 
It will be helpful to run with verbose flag enabled (see the README).

**Comment by sbmpost** [comment]
@fredngo 
> Launchpad is buggy when the autoraise is running. I'll report more as I observe the behavior, if I can.

What settings do you use when running AutoRaise? Please use the command line version and explicitly pass the parameters you are using. Paste the output here, so something like:
    v3.0 by sbmpost(c) 2022, usage:

    AutoRaise
      -delay <0=no-raise, 1=no-delay, 2=50ms, 3=100ms, ...>
      -warpX <0.5> -warpY <0.5> -scale <2.0>
      -mouseStop <true|false>
      -verbose <true|false>

    Started with:
      * delay: 0ms
      * mouseStop: false
      * verbose: true

**Comment by git-rz** [comment]
I can reproduce it. Here is my output:

```
v3.0 by sbmpost(c) 2022, usage:

AutoRaise
  -delay <0=no-raise, 1=no-delay, 2=50ms, 3=100ms, ...>
  -warpX <0.5> -warpY <0.5> -scale <2.0>
  -mouseStop <true|false>
  -verbose <true|false>

Started with:
  * warp and focus only (no-raise)
  * warpX: 0.5, warpY: 0.5, scale: 2.0
  * mouseStop: true
  * verbose: false
 ```

Issue does not happen for me:
- as long as the mouse is in motion
- if I don't move the mouse after activating launchpad in a hot corner

I hadn't noticed this before as my normal workflow is to just use the keyboard to find the needed app.

**Comment by dreeves** [comment]
New bug report with focus-follows-mouse.

I can't click on this volume menu:

<img width="248" alt="image" src="https://user-images.githubusercontent.com/9928/167986548-e3c2644e-883f-4a8d-ba13-d69d96ca264a.png">

The menu instantly goes away when I move the mouse. Same with the bluetooth and battery and wifi menus, but other menus are fine.

**Comment by dreeves** [comment]
PPS: I don't think I like mouseStop=true :( I seem to sometimes switch windows with my thumb on the trackpad and try to start typing before the mouse has stopped. I'll keep experimenting though, to see how much of a problem that really is in practice!

**Comment by dreeves** [comment]
Confirmed that I'm on the latest of the `7-47-focus-without-raise-experimental` branch (commit `7ce12e677e874c4745a8abdbd9daab7b0b4105e0`). Here's the startup output:

```
v3.0 by sbmpost(c) 2022, usage:

AutoRaise
  -delay <0=no-raise, 1=no-delay, 2=50ms, 3=100ms, ...>
  -warpX <0.5> -warpY <0.5> -scale <2.0>
  -mouseStop <true|false>
  -verbose <true|false>

Started with:
  * delay: 500ms
  * mouseStop: false
  * verbose: false
```

I think my menu bar just looks weird because of how I cropped the screenshot.

And, again, setting mouseStop=true fixes it but makes it seem sluggish. (I'm still uncertain if it's really a problem in practice to have mouseStop=true.)

**Comment by sbmpost** [comment]
@dreeves 
I am going to upgrade to latest Monterey. In the meantime, can you edit the line in AutoRaise.mm where it says:
    
    #define MENUBAR_CORRECTION 6

and try a higher value? 10 or something?

**Comment by sbmpost** [comment]
@dreeves 
I have upgraded and I also pushed a fix.

**Comment by dreeves** [comment]
Fix confirmed! I can now use the menu bar without needing the mouseStop=true option. Thanks so much!

**Comment by fredngo** [comment]
> @git-rz
> 
> > I can reproduce it.
> 
> I can't reproduce this either. I must be missing something, but can you double check that you are using the latest code?

Sorry smbpost, he says he **can** reproduce it...

I haven't had a chance to try though, just completely slammed right now. Maybe this weekend.

**Comment by git-rz** [comment]
To confirm, I can still reproduce the issue where Lauchpad goes away if the mouse stops moving. I am using the latest code:
```
v3.1 by sbmpost(c) 2022, usage:

AutoRaise
  -delay <0=no-raise, 1=no-delay, 2=50ms, 3=100ms, ...>
  -warpX <0.5> -warpY <0.5> -scale <2.0>
  -mouseStop <true|false>
  -verbose <true|false>

Started with:
  * warp and focus only (no-raise)
  * warpX: 0.5, warpY: 0.5, scale: 2.0
  * mouseStop: true
  * verbose: false
```

**Comment by git-rz** [comment]
https://user-images.githubusercontent.com/53906153/168412906-a05a87ee-4052-40ae-a12c-502e2b6b2964.mov

^^ shows the issue.

**Comment by fredngo** [comment]
> ^^ shows the issue.

Thank you so much @git-rz ! I don't think I can do any better in reproducing the problem!

**Comment by sbmpost** [comment]
@git-rz 
Thanks for the movie. I can reproduce this on my machine now. Although launchpad looks like a standard window, it seems to be more like an overlay. Therefore moving the mouse over the overlay, will simply find the windows below it, and if AutoRaise decides one of these windows needs to be focused/raised, the overlay (launchpad) is closed. I am not sure this is easy to solve because:

    1. The overlay is never considered to be 'under' the mouse, as far as AutoRaise is concerned it is not there.
    2. There is no way that I know of to see if the launchpad is active. There are no events for it.

Maybe it is possible to find the launchpad overlay by iterating all UI elements, but that would be very CPU intensive. See also these links:

https://forums.macrumors.com/threads/how-to-tell-if-launchpad-is-open.1568658/

https://stackoverflow.com/questions/10359558/osx-lion-nsevent-when-launchpad-is-entered

**Comment by fredngo** [comment]
As I understand it then, this launchpad problem may not be with the experimental branch?

**Comment by git-rz** [comment]
@sbmpost - An idea to work-around the inability to detect the overlay:

If mouse travels all the way to a lauchpad-opening corner (or any corner if you can't detect hotcorner config), then  suspend all autoraise / autofocus until a launchpad-closing event happens, such as a click, `esc` or `enter`.

**Comment by fredngo** [comment]
> If mouse travels all the way to a lauchpad-opening corner (or any corner if you can't detect hotcorner config), then suspend all autoraise / autofocus until a launchpad-closing event happens, such as a click, `esc` or `enter`.

Great idea!

Also, perhaps best (if possible) to generalize detecting if Launchpad opens in any fashion, via hot corners or, like me, actually launching it using the icon... Am I weird? 😅

<img width="526" alt="Screen Shot 2022-05-14 at 7 42 44 PM" src="https://user-images.githubusercontent.com/11791/168451481-5c903253-e147-4b4e-8f85-3763891e8e1d.png">

**Comment by sbmpost** [comment]
@fredngo @git-rz 
I created some magic which just might do the trick. But for every piece of magic, there is a price to pay. I expect AutoRaise to be slower/cpu to increase as I am iterating UI elements of the Dock app. Perhaps it is not very much noticable, but I will leave that decision to you guys ;-). Just pull latest code, and the launchpad problem should be gone.

**Comment by git-rz** [comment]
The new code is working wonderfully. The price is not perceptible. Nor is it seen in the CPU monitor. Really, there is almost no usage at all from Autoraise.

**Comment by sbmpost** [comment]
@git-rz 
Thanks for these numbers.
I see similar results on my machine. Not sure about older ones though. And good that it works of course :)

**Comment by git-rz** [comment]
Working great so far:

<img width="873" alt="image" src="https://user-images.githubusercontent.com/53906153/168730856-2ff8b3df-5270-42ed-9a8e-7ae8d4df85b2.png">

**Comment by fredngo** [comment]
This is awesome, guys.

I've updated my homebrew formula to compile with the `EXPERIMENTAL_FOCUS_FIRST` flag, so you can:

```zsh
brew tap fredngo/autoraise-experimental
brew install --HEAD fredngo/autoraise-experimental/autoraise
brew services start autoraise
```

and it'll automatically compile the binary with that flag. I'm still using the experimental branch in the homebrew formula, but at some point can switch it to master if the branch goes away.

My `~/.config/AutoRaise/config` file has the following:

```
delay=0
mouseStop=true
```

In preliminary usage it seems to work very well for me!

* I have focus but no raise
* I can select the menu bar even if traversing across multiple other windows
* Launchpad doesn't disappear anymore

Amazing work @sbmpost and everyone on this thread!

**Comment by dreeves** [comment]
Yeah, I'm pretty blown away by @sbmpost's work on this. And a wonderful community forming here! Thanks, y'all!

Getting it all into master is really nice too. I don't know if it's realistic to have immediate-focus aka focus-first be a setting rather than a compiler flag or even just How It Works. In theory I don't think delayed focus ever makes sense (only delayed autoraise) but obviously I don't understand the ways Apple is making that fraught or fragile. I'm really grateful to have it working exactly how I want now, to the point of avoiding updating macOS like the plague in case Apple breaks this.

Naive idea: instead of a compiler flag, what if it's a compile-time check that enables focus-first if you're on any of an enumerated list of known-compatible versions of macOS?

**Comment by fredngo** [comment]
Ya personally I'd definitely prefer EXPERIMENTAL_FOCUS to be a config flag instead of a compile-time flag so I can put something like `experimentalFocus=true` in the config file. Then, I wouldn't need my homebrew formula anymore. But that's up to @smbpost!

**Comment by sbmpost** [comment]
Thanks for all your nice comments and happy that it works well for all of you. @dreeves: indeed it almost feels like a community ;-)

About the option vs compile time flag: we might want to change that in the future, but I wanted to get this to master quickly. A compile flag is also slightly more efficient in terms of performance of course. But not a deal breaker to make it an option someday. My time is now more limited, so let's hope this version can withstand the future for a while. Most likely I will close this issue sometime next week if nothing else pops up.

Thanks everyone once more! 😃

**Comment by sbmpost** [comment]
@fredngo @Laaarge @samholmes @dreeves @git-rz @BenJackGill 
I actually had to push a v3.1 to master because I noticed some small (barely notable) regressions.

**Comment by fredngo** [comment]
> I actually had to push a v3.1 to master because I noticed some small (barely noticeable) regressions.

I use bartender also so I guess I better upgrade to that... I've updated my homebrew formula to use master now. I guess I'll have to open a pull request to the original homebrew formula maintainer at this point and get rid of my fork.

**Comment by fredngo** [comment]
FYI I've added the relevant commits to the original homebrew formula; whenever the [pull request](https://github.com/Dimentium/homebrew-autoraise/pull/2) is done, my fork of the homebrew formula will become obsolete

**Comment by samholmes** [comment]
I built AutoRaise from source. Now there's a build on homebrew? What are the exact steps for reference to use the new feature?

**Comment by fredngo** [comment]
> I built AutoRaise from source. Now there's a build on homebrew? What are the exact steps for reference to use the new feature with all the fixes?

Well, the pull request hasn't been accepted into the original homebrew formula yet, but you can check out the documentation on [my fork](https://github.com/fredngo/homebrew-autoraise-experimental)

**Comment by samholmes** [comment]
@fredngo I get an error trying to install from homebrew:

```
❯ brew install autoraise --with-dexperimental_focus_first
...
Error: invalid option: --with-dexperimental_focus_first
```

**Comment by samholmes** [comment]
```
❯ brew info autoraise
dimentium/autoraise/autoraise: stable 2.9, HEAD
Focus-follows-mouse for MacOS
https://github.com/sbmpost/AutoRaise
Not installed
From: https://github.com/dimentium/homebrew-autoraise/blob/HEAD/autoraise.rb
License: GPL-3.0
==> Options
--with-dalternative_task_switcher
	CXXFLAGS=-DALTERNATIVE_TASK_SWITCHER
--HEAD
	Install HEAD version
==> Caveats
To restart dimentium/autoraise/autoraise after an upgrade:
  brew services restart dimentium/autoraise/autoraise
Or, if you don't want/need a background service you can just run:
  AutoRaise
```

Opening https://github.com/dimentium/homebrew-autoraise/blob/HEAD/autoraise.rb#L9 and you see there is only one option available,

**Comment by fredngo** [comment]
> > I built AutoRaise from source. Now there's a build on homebrew? What are the exact steps for reference to use the new feature with all the fixes?
> 
> Well, the pull request hasn't been accepted into the original homebrew formula yet, but you can check out the documentation on [my fork](https://github.com/fredngo/homebrew-autoraise-experimental)

As I said, my pull request is still pending

**Comment by git-rz** [comment]
@sbmpost, I am seeing another issue, however I'm not sure if AutoRaise or iTerm2 is the best place for the fix.

I am using focus-follows-mouse support within iTerm2. I expect the pane underneath the mouse pointer to focus when the mouse moves to the pane.

The issue is that when using Autoraise in Focus-only mode, and when the pointer returns to a non-active pane of iTerm2 from a window of a different application, then the pane does not get the focus.

The pane only gets focus when the pointer moves across a pane boundary within iTerm2.

To reproduce: split  iTerm2 into two panes, top and bottom. Snap the iTerm window to the right half of the screen. Snap a browser to the left half. Position the mouse over top half of iterm. Mouse left over the browser. Then reenter iterm on the bottom half.

Observe that the top half of iTerm is still activated.

**Comment by sbmpost** [comment]
@git-rz 
> The issue is that when using Autoraise in Focus-only mode, and when the pointer returns to a non-active pane of iTerm2 from a window of a different application, then the pane does not get the focus.

It looks more like an iTerm bug actually. Try the following to see what I mean:

1. Disable AutoRaise (stop it)
2. Split your iTerm and arrange a browser on the left
3. Focus your iTerm by clicking in the top pane
4. Move the mosue out of the iTerm to the browser
5. Click the browser to give it the focus
6. Move the mouse to the bottom iTerm pane
7. Focus iTerm with the OSX task switcher (using cmd-tab)

After step 7, I would expect the bottom iTerm pane to be focused, but it doesn't happen. iTerm should check if the correct pane is focused not only when the mouse clicks it, but also if it receives the focus via other means (a keyboard shortcut or AutoRaise).

**Comment by git-rz** [comment]
@smbpost, indeed, I suspected the same.

However, in the steps to reproduce, given that only keyboard navigation is involved, in that case I would argue iTerm should NOT refocus the pane. 

Consider the following:
- iTerm window, full screen, split pane.
- Browser window, full screen.
- User editing notes in iTerm.
- User cmd-tab to browser, mouse moves within browser such that another pane would be activated if iTerm were on top
- User cmd-tab back to iTerm to update notes.

In this case, the user would again be annoyed that the focus moved. Even though the mouse is over the pane, the user didn't actually move the mouse within iTerm, so it should not be considered. I will report the behavior to iTerm, but will need to formulate in a way that may actually get implemented.

Unless perfect interoperability with an experimental build of AutoRaise is something the devs over there would consider, I would need to come up with another set of steps that can recreate the issue.

**Comment by sbmpost** [comment]
@Laaarge @dreeves @fredngo @samholmes @BenJackGill @git-rz 

So far it seems all is well with release 3.1. It has been quite a ride to solve this issue. I am closing it now, thank you all for reporting and additional feedback.

**Comment by samholmes** [comment]
I’d love it if I can click to raise without it interacting with the window
on initial click. But, that’s just me being spoiled at this point.

**Comment by swissbuechi** [comment]
Can someone confirm this experimental feature is still working on Ventura 13.1?
For me it seems like compiling with `make CXXFLAGS="-DEXPERIMENTAL_FOCUS_FIRST" && make install` and adding `delay=0` to my `~/.config/AutoRaise/config` is having no effect on the auto focus.
When i set `delay=1`, the auto focus + auto rise work just fine (still compiled with the experimental focus first feature).

Thanks for any feedback.

**Comment by elaw-fastly** [comment]
I would say that this works quite well on macOS 13.1 Ventura
I'm on a MBP with M1 silicon. 
I've been desperately wanting this feature since 2015!

Focus clearly moves to chrome on mouse-enter, and the keyboard focus is usually returned to the last element keyboard focus was on (usually the location bar or input field) **for the window I'm hovering over.** (nice!)

There are some _small_ quirks
* hovering over iterm2 (with focus-follows-mouse enabled) works as expected, but the iterm2 window is occasionally raised if the cursor was previously over the system notification popup or from abother app _on a separate display_ - this is a very minor nit.
* iterm2 preferences popup loses keyboard focus as you move away from it into the iterm2 window (as expected), but moving back over the preferences causes the iterm2 window to raise.  Not a big deal for me as cmd+, brings it right back.
* warping occurred only once during my testing (I could tell from the cursor jump and the scaling effect), but didn't always trigger
* I had some kind of temporal edge case where moving from iterm2 to slack didn't focus on slack, but when I transitions from chrome to slack it worked.  Clicking windows to change layering order or moving them seemed to make it work again and I cannot recreate.

**Comment by ericslaw** [comment]
I would say that this works quite well on macOS 13.1 Ventura
I'm on a MBP with M1 silicon. 
I've been desperately wanting this feature since 2015!

Focus clearly moves to chrome on mouse-enter, and the keyboard focus is usually returned to the last element keyboard focus was on (usually the location bar or input field) **for the window I'm hovering over.** (nice!)

There are some _small_ quirks
* hovering over iterm2 (with focus-follows-mouse enabled) works as expected, but the iterm2 window is occasionally raised if the cursor was previously over the system notification popup or from abother app _on a separate display_ - this is a very minor nit.
* iterm2 preferences popup loses keyboard focus as you move away from it into the iterm2 window (as expected), but moving back over the preferences causes the iterm2 window to raise.  Not a big deal for me as cmd+, brings it right back.
* warping occurred only once during my testing (I could tell from the cursor jump and the scaling effect), but didn't always trigger
* I had some kind of temporal edge case where moving from iterm2 to slack didn't focus on slack, but when I transitions from chrome to slack it worked.  Clicking windows to change layering order or moving them seemed to make it work again and I cannot recreate.

**Comment by sbmpost** [comment]
@ericslaw 
Thanks for this writeup :-)

I noticed you use a very large value for delay which effectively disables it. You can achieve the same thing with:

    -focusDelay 1 -delay 0

The warp feature might not always work consistently if the alternative task switcher is enabled. It depends a bit on how you do the actual task switching. Using only cmd-tab without alternative task switcher works reliably.

**Comment by elaw-fastly** [comment]
My initial experiments using -delay 0 shows functionality was inconsistent. I have now set it to 0 and it is working as designed. I attribute this to the iTerm2 focus-follows-mouse interfering with AutoRaise 
and have since disabled that feature of iTerm2.

**Comment by ericslaw** [comment]
My initial experiments using -delay 0 shows functionality was inconsistent. I have now set it to 0 and it is working as designed. I attribute this to the iTerm2 focus-follows-mouse interfering with AutoRaise
and have since disabled that feature of iTerm2.

**Comment by paulmil1** [comment]
@sbmpost   
can the EXPERIMENTAL_FOCUS_FIRST finally become a config flag? i'd like to test some configurations to see which works best for me, but its impossible without messy recompiling

**Comment by sbmpost** [comment]
@paulmil1 
Sorry but this is probably never going to be an option because it relies on a private Apple API which can change and break this functionality. This feature therefore isn't "officially" supported and will have to remain experimental until Apple provides a way to focus only instead of raising

**Comment by sbmpost** [comment]
@paulmil1 
Btw to test out these configurations, compiling just once should do right? After that the other settings don't require recompilation. Also the GUI version (provided as a dmg, see the README) allows you to quickly change those settings. The GUI version was compiled with EXPERIMENTAL_FOCUS_FIRST enabled.

**Comment by henrik242** [comment]
I can confirm that focus-without-raise works with `brew install dimentium/autoraise/autoraise --with-dexperimental_focus_first`,  an `~/.Autoraise` with `delay=0` and `brew services start autoraise`.  Great stuff!

**Comment by dreeves** [comment]
Thanks @henrik242! Is that better or different than my existing workflow:

```
git clone https://github.com/sbmpost/AutoRaise.git autoraise
cd autoraise
make CXXFLAGS="-DEXPERIMENTAL_FOCUS_FIRST"
./AutoRaise -pollMillis 50 -delay 11 -focusDelay 1
```

PS: And I just want to re-pitch my proposal to refactor the -delay parameter so "0" means "no delay": https://github.com/sbmpost/AutoRaise/issues/47#issuecomment-1115102265

**Comment by crash48** [comment]
I'm a noob in MacOS and homebrew, but I got this installed and working. However what annoys me now, is that if I have a window somewhere on the right and want to access the file menu and go to ex. settings for that app, if I accidentally hover over another window it changes the options on my menu bar. Because I installed the homebrew version, I have no button to toggle autoraise in my top menu bar. Is there a way to get the experimental focus option AND the icon to toggle it on and off?

**Comment by norrbru** [comment]
Hi! This is the exact sort of behaviour that I've been chasing for a long time. 
Any updates on functionality now with MacOS Tahoe? Would this still work, are there any new possible workarounds?
